pkg/tf2pulumi/convert/model_helpers.go

```diff
@@ -0,0 +1,81 @@
+// Copyright 2016-2021, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// This module provides helper functions to work with
+// `github.com/pulumi/pulumi/pkg/v3/codegen/hcl2/model` that have not
+// yet made it upstream.
+
+package convert
+
+import (
+	"fmt"
+	"github.com/hashicorp/hcl/v2"
+	//"strings"
+
+	"github.com/pulumi/pulumi/pkg/v3/codegen/hcl2/model"
+)
+
+// Helps assign a type to a `Type="config"` block. It is currently not
+// apparent from the types, but blocks carry 1 or 2 labels, with the
+// second optional label encoding the type of the variable.
+//
+// Example block in HCL syntax:
+//
+//     config foo "int" {
+//        default = 1
+//     }
+//
+// The type is decoded downstream (as in
+// `github.com/pulumi/pulumi/pkg/codegen/pcl/binder.go`) in PCL
+// processing. Currently we cannot guarantee that all posible
+// `model.Type` values survive the string encoding and subsequent
+// decoding. Thereore this function mimicks downstream decoding to see
+// if the value turns around, and aborts processing if that is not the
+// case.
```

Comment:
Just to clarify: if we remove the explicit type from this example, is the issue that the type of `default` is `const(1)` and the binder then infers the type of `foo` to be `const(1)`? If so, it seems like we ought to fix this in the PCL binder s.t. the type of `foo` is inferred to be the const type's underlying type.

Reply:
@pgavlin can we discuss the surrounding code here?

In a sense, you are right, there is an issue with inferring Const types for a variable from a `default` value.  These can (and are) nested. In the failing example it's nested under `ObjectType`:

```
	constType := model.NewConstType(model.IntType, cty.NumberIntVal(1))
	checkError("constType", constType)

	objTypeWithConst := model.NewObjectType(map[string]model.Type{"foo": constType})
	checkError("objectTypeWithConstType", objTypeWithConst)
```

We could insert a transformation `func unConstify(t model.Type) model.Type` and use it in here, and it *may* fix the manifested issue. Would it be possible to backlog it and check the fix as is as it is on the critical path? I'd like to take the time on that to fully understand the implications and recur through the model.Type grammar.

Here are some of the things that are quite surprising to me that make me tread really cautiously:

1. Why does the system rely on turning model.Type into a string and then parsing it back? We know it does not work if model.Type tree includes a Const node. Will it work for all possible cases if we remove Const nodes? Can we test that? I do not yet trust that will work, therefore aggressively checking that it will work or backing off to Dynamic type seems warranted.

2. How does this even work, why does `model.BindExpressionText` parse a string formed by taking `model.Type` through a `Sprintf("%v")` into an expression with a type equivalent to the original type? I am looking at this doc string `// BindExpressionText parses and binds an HCL2 expression using the given scope.` and it seems extremely surprising. Is the type expression grammar embedded in the term expression grammar? I think there is a bit of a documentation gap here at the least, we can correct.

LMK what you think, I still lean to merging as is, backlogging type inference from default value to remove consts, then also backlogging testing and documenting model.Type turnaround through a string and removing the model_helpers code once we can trust that to work.

Reply:
I'm doing a little exploration here, I think I recovered the grammar of model.Type so I can get some answers to my own questions shortly.

Reply:
https://github.com/pulumi/pulumi-terraform-bridge/pull/410 - not only Const types do not turnaround; there's a lot more to it.... The current aggressive check should stay in I think for now.  Let me explore deconstifying the types.